synapse/storage/databases/main/events_worker.py

```diff
@@ -1626,3 +1626,136 @@ def _cleanup_old_transaction_ids_txn(txn):
             "_cleanup_old_transaction_ids",
             _cleanup_old_transaction_ids_txn,
         )
+
+    async def event_next_to_gap(self, room_id: str, event_id: str) -> bool:
+        """Check if the given `event_id` is  next to a gap of missing events.
+        Looks at gaps going forwards and backwards. The gap in front of the
+        latest events is not considered a gap.
+
+        Args:
+            room_id: room where the event lives
+            event_id: event to check
+
+        Returns:
+            Boolean indicating whether it's an extremity
+        """
+
+        def event_next_to_gap_txn(txn) -> bool:
+            # If the event in question is listed as a backward extremity, it's
+            # next to a gap.
+            backward_extremity_query = """
+                SELECT 1 FROM event_backward_extremities
+                LEFT JOIN rejections USING (event_id)
+                WHERE
+                    room_id = ?
+                    AND event_id = ?
+                    /* Make sure event is not rejected */
+                    AND rejections.event_id IS NULL
+                LIMIT 1
+            """
+
+            # Check to see whether the event in question is already referenced
+            # by another event. If we don't see any edges, we're next to a
+            # forward gap.
+            #
+            # We can't just check `event_forward_extremities` directly because that
+            # doesn't include potential backfilled and outlier events.
+            #
+            # We consider any event that is an `event_forward_extremities` as
+            # the latest in the room and not next to a gap.
+            forward_edge_query = """
+                SELECT 1 FROM event_edges
+                LEFT JOIN rejections ON event_edges.prev_event_id == rejections.event_id
+                LEFT JOIN event_forward_extremities ON event_edges.prev_event_id == event_forward_extremities.event_id
+                WHERE
+                    event_edges.room_id = ?
+                    AND event_edges.prev_event_id = ?
+                    /* If the event in question is a forward extremity, we will
+                     * just consider the forward gap as not a gap since it's one
+                     *  of the latest events in the room.
+                     */
+                    AND event_forward_extremities.event_id IS NOT NULL
```

Comment:
If `event_id` has no forward edges and is in `event_forward_extremities`, the query will return no rows and we'll still class the event as next to a gap. It might be easier to do a separate query for `event_forward_extremities`.

Reply:
Updated and explained this edge case in the comments 👍